File: modules/09-computer-vision/ts/task1_classify.ts
/**
 * task1_classify.ts — image classification with a pretrained model.  🟢
 *
 * What it teaches:
 *   A pretrained ViT (Vision Transformer) model assigns ImageNet labels to an
 *   image. We use @huggingface/transformers (transformers.js) so this runs
 *   entirely in Node — no Python, no GPU, no server required. The model is
 *   downloaded once (~90 MB for the quantised version) and cached locally.
 *
 *   If you prefer NOT to download the model, use the HOSTED path below which
 *   calls the HuggingFace Inference API instead (requires HF_TOKEN).
 *
 * How to run (from the repo root):
 *   pnpm tsx modules/09-computer-vision/ts/task1_classify.ts
 *
 *   # Force the hosted API path:
 *   USE_HF_API=true pnpm tsx modules/09-computer-vision/ts/task1_classify.ts
 *
 * Environment variables (in .env):
 *   HF_TOKEN     — required only when USE_HF_API=true
 *   USE_HF_API   — set to "true" to skip local model download
 *
 * Acceptance:
 *   - Prints a ranked list of label/score pairs for the sample image.
 *   - Works via local transformers.js (default) or hosted API.
 */

import "dotenv/config";
import { readFileSync, existsSync } from "node:fs";
import { mkdir } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import https from "node:https";
import fs from "node:fs";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ASSETS_DIR = path.join(__dirname, "..", "assets");
const SAMPLE_IMAGE = path.join(ASSETS_DIR, "cat.jpg");

// ViT-base pretrained on ImageNet (1000 classes).
// transformers.js will download + cache this on first run.
const HF_MODEL = "Xenova/vit-base-patch16-224";

interface ClassificationResult {
  label: string;
  score: number;
}

// ---------------------------------------------------------------------------
// LOCAL path — transformers.js runs the model in Node (default)
// ---------------------------------------------------------------------------

async function classifyLocal(imagePath: string): Promise<ClassificationResult[]> {
  // TODO (exercise): run the model locally with transformers.js.
  //   - Import `pipeline` from "@huggingface/transformers" and `await` it to
  //     build an "image-classification" pipeline for HF_MODEL.
  //   - Call the classifier with `imagePath` (the pipeline accepts a file path
  //     or URL string). It resolves to a { label; score }[] array.
  //   - Return that array sorted by score, highest first.

  throw new Error(
    "Complete the TODO: import pipeline from @huggingface/transformers, " +
    "create a classifier, and call it with imagePath."
  );
}

// ---------------------------------------------------------------------------
// HOSTED path — HuggingFace Inference API (no local model download)
// ---------------------------------------------------------------------------

async function classifyHosted(imagePath: string): Promise<ClassificationResult[]> {
  const token = process.env.HF_TOKEN;
  if (!token) {
    throw new Error(
      "Missing HF_TOKEN. Get a free token at huggingface.co/settings/tokens " +
      "and add HF_TOKEN=... to your .env file."
    );
  }

  const imageBytes = readFileSync(imagePath);
  const model = "google/vit-base-patch16-224";

  // TODO (exercise): call the HF Inference API endpoint:
  //   POST https://api-inference.huggingface.co/models/<model>
  //   Headers: Authorization: Bearer <token>, Content-Type: application/octet-stream
  //   Body: raw image bytes
  //
  // The response is JSON: [{ "label": "...", "score": 0.9 }, ...]
  //
  // Hint: use the built-in `fetch` (Node 18+) or `node:https`.

  const response = await fetch(
    `https://api-inference.huggingface.co/models/${model}`,
    {
      method: "POST",
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/octet-stream",
      },
      body: imageBytes,
    }
  );

  if (!response.ok) {
    const body = await response.text();
    throw new Error(`HF API error ${response.status}: ${body}`);
  }

  // TODO (exercise): the JSON response is already parsed into `data` (a
  // ClassificationResult[]). Return it sorted by score descending, then remove
  // the throw below.
  const data = (await response.json()) as ClassificationResult[];

  throw new Error(
    "Parse the response JSON and return the results sorted by score descending."
  );
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function ensureSampleImage(): Promise<string> {
  await mkdir(ASSETS_DIR, { recursive: true });
  if (existsSync(SAMPLE_IMAGE)) return SAMPLE_IMAGE;

  const url =
    "https://upload.wikimedia.org/wikipedia/commons/thumb/4/4d/Cat_November_2010-1a.jpg/320px-Cat_November_2010-1a.jpg";
  console.log(`Downloading sample image → ${SAMPLE_IMAGE}`);

  await new Promise<void>((resolve, reject) => {
    const file = fs.createWriteStream(SAMPLE_IMAGE);
    https.get(url, (res) => {
      res.pipe(file);
      file.on("finish", () => { file.close(); resolve(); });
    }).on("error", reject);
  });

  return SAMPLE_IMAGE;
}

async function main(): Promise<void> {
  const imagePath = await ensureSampleImage();
  const useApi = process.env.USE_HF_API === "true";

  console.log(`Image    : ${path.basename(imagePath)}`);
  console.log(`Mode     : ${useApi ? "HF Inference API (hosted)" : "transformers.js (local)"}`);
  console.log(`Model    : ${useApi ? "google/vit-base-patch16-224" : HF_MODEL}`);
  console.log();

  const results = useApi
    ? await classifyHosted(imagePath)
    : await classifyLocal(imagePath);

  console.log("Top predictions:");
  results.slice(0, 5).forEach((item, i) => {
    console.log(`  ${i + 1}. ${item.label.padEnd(40)} ${item.score.toFixed(3)}`);
  });
}

main().catch((err) => {
  console.error(err.message ?? err);
  process.exit(1);
});

File: modules/09-computer-vision/py/task1_classify.py
"""task1_classify.py — image classification with a pretrained model.  🟢

What it teaches:
    How a pretrained vision model assigns a label (or ranked list of labels) to
    an image. We use the HuggingFace Inference API so no GPU or multi-GB model
    download is needed — the model lives in HF's cloud.

How to run (from the repo root):
    uv run python modules/09-computer-vision/py/task1_classify.py

Environment variables (in .env):
    HF_TOKEN   — HuggingFace token (get one free at huggingface.co/settings/tokens)

Optional local path (skip the API, run model on your machine):
    uv sync --extra vision         # installs torch, torchvision, transformers, pillow
    Then uncomment the LOCAL block below and comment out the HOSTED block.
    Warning: first run downloads ~350 MB of model weights; slow on Mac MPS.

Acceptance:
    - Prints a ranked list of label/score pairs for the sample image.
    - You can swap IMAGE_PATH to any local JPEG/PNG and it still works.
"""

from __future__ import annotations

import os
from pathlib import Path

from dotenv import load_dotenv

load_dotenv()

# Path to a sample image — uses the bundled one or any local file you provide.
SAMPLE_IMAGE = Path(__file__).parent.parent / "assets" / "cat.jpg"

# HuggingFace model for image classification (ImageNet-1k, 1000 classes).
# See: https://huggingface.co/google/vit-base-patch16-224
HF_MODEL = "google/vit-base-patch16-224"


# ---------------------------------------------------------------------------
# HOSTED path — HuggingFace Inference API (default, no local GPU needed)
# ---------------------------------------------------------------------------

def classify_hosted(image_path: Path) -> list[dict]:
    """Call HF Inference API to classify the image. Returns list of
    {'label': str, 'score': float} dicts sorted by score descending."""
    try:
        from huggingface_hub import InferenceClient
    except ImportError:
        raise ImportError(
            "huggingface_hub is not installed. Run: uv sync --extra vision\n"
            "Or install manually: pip install huggingface-hub"
        )

    token = os.getenv("HF_TOKEN")
    if not token:
        raise RuntimeError(
            "Missing HF_TOKEN. Get a free token at huggingface.co/settings/tokens "
            "and add HF_TOKEN=... to your .env file."
        )

    client = InferenceClient(token=token)

    # The InferenceClient accepts a file path, URL, or raw bytes.
    with open(image_path, "rb") as f:
        image_bytes = f.read()

    # TODO (exercise): classify the image via the client.
    #   - Call `client.image_classification(...)`, passing the raw `image_bytes`
    #     and telling it which `model` to use (HF_MODEL).
    #   - The return value is a list of ClassificationOutput objects, each with
    #     `.label` and `.score` attributes.
    results = None  # replace with the API call

    raise NotImplementedError(
        "Complete the TODO above: call client.image_classification() "
        "and assign the result to `results`."
    )

    # TODO (exercise): return `results` as a plain list of `{"label": ..., "score": ...}`
    # dicts (read the .label / .score off each item) so they print cleanly.
    return results  # type: ignore[return-value]


# ---------------------------------------------------------------------------
# LOCAL path (optional) — transformers + PIL, runs entirely offline
# ---------------------------------------------------------------------------

def classify_local(image_path: Path) -> list[dict]:
    """Classify using a locally-downloaded ViT model via transformers.
    Uncomment and call this instead of classify_hosted() to run offline.
    First run will download ~350 MB of model weights.
    """
    # TODO (optional exercise): implement the offline path.
    #   - From `transformers`, build an "image-classification" `pipeline(...)`
    #     pointed at HF_MODEL; open the file with PIL's `Image.open(...)` and
    #     convert it to "RGB".
    #   - Run the pipeline on the image (it returns a list of
    #     `{"label": ..., "score": ...}` dicts) and return them sorted by score,
    #     highest first.

    raise NotImplementedError("Local path is not implemented yet — see the TODO above.")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> None:
    image_path = SAMPLE_IMAGE
    if not image_path.exists():
        # Fallback: download the sample image automatically on first run.
        image_path = _download_sample_image()

    print(f"Image    : {image_path.name}")
    print(f"Model    : {HF_MODEL} (via HF Inference API)")
    print()

    results = classify_hosted(image_path)

    print("Top predictions:")
    for i, item in enumerate(results[:5], 1):
        label = item["label"] if isinstance(item, dict) else item.label
        score = item["score"] if isinstance(item, dict) else item.score
        print(f"  {i}. {label:<40} {score:.3f}")


def _download_sample_image() -> Path:
    """Download a royalty-free sample image on first run."""
    import urllib.request

    assets_dir = Path(__file__).parent.parent / "assets"
    assets_dir.mkdir(parents=True, exist_ok=True)
    dest = assets_dir / "cat.jpg"

    # Wikimedia Commons — public domain cat photo.
    url = "https://upload.wikimedia.org/wikipedia/commons/thumb/4/4d/Cat_November_2010-1a.jpg/320px-Cat_November_2010-1a.jpg"
    print(f"Downloading sample image from Wikimedia Commons → {dest}")
    urllib.request.urlretrieve(url, dest)  # noqa: S310
    return dest


if __name__ == "__main__":
    main()
